ocaml/typing/typemod.ml

```diff
@@ -3307,6 +3314,37 @@ let gen_annot outputprefix sourcefile annots =
   Cmt2annot.gen_annot (Some (outputprefix ^ ".annot"))
     ~sourcefile:(Some sourcefile) ~use_summaries:false annots
 
+let check_argument_type_if_given env sourcefile actual_sig arg_module_opt =
+  match arg_module_opt with
+  | None -> None
+  | Some arg_module ->
+      let arg_import =
+        (* This will soon be converting from one type to another *)
+        arg_module
+      in
+      Env.register_parameter_import arg_import;
```

Comment:
I'm not quite sure what this is doing. It seems like it would just mark `arg_import` as being a parameter, but I would have thought that would happen automatically when the `cmi` was loaded.

Reply:
It does just mark it as a parameter. We don't it do it automatically because currently the only place where we enforce the rule “no referring to a parameter without `-parameter`” is when a .cmi is loaded: we complain anytime we load a parameter and didn't have a specific reason to expect one. Now that I'm explaining it, it doesn't seem ideal.

Reply:
IIRC there are different code paths for "Please read this specific cmi file" and "Please find identifier `Foo`". Probably you can just only error on the second one. I think you've probably stratified the env enough that you can associate the error with the layer that works in terms of globals rather than the layer that works in terms of cmi files. That should let you set thinks up so that even if you read a specific cmi and then asked for its identifier it would successfully error on the second call.

Reply:
Indeed, I found there was already a CR-someday in `persistent_env.ml` to move the check to the right place. Pushed a change that gets rid of this “pre-register things that we expect to be parameter modules” logic, so this call to `Env.register_parameter_import` is simply no longer there.